// sigma npm/rendering/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EdgeTriangleProgram = exports.EdgeRectangleProgram = exports.EdgeLineProgram = exports.EdgeClampedProgram = exports.EdgeArrowHeadProgram = exports.EdgeArrowProgram = exports.NodePointProgram = exports.NodeCircleProgram = exports.EdgeProgram = exports.NodeProgram = void 0;
/**
 * Just some shortcuts to ease importing various rendering related things from
 * the codebase:
 */
var node_1 = require("./node");
Object.defineProperty(exports, "NodeProgram", { enumerable: true, get: function () { return node_1.NodeProgram; } });
var edge_1 = require("./edge");
Object.defineProperty(exports, "EdgeProgram", { enumerable: true, get: function () { return edge_1.EdgeProgram; } });
var node_circle_1 = require("./programs/node-circle");
Object.defineProperty(exports, "NodeCircleProgram", { enumerable: true, get: function () { return __importDefault(node_circle_1).default; } });
var node_point_1 = require("./programs/node-point");
Object.defineProperty(exports, "NodePointProgram", { enumerable: true, get: function () { return __importDefault(node_point_1).default; } });
var edge_arrow_1 = require("./programs/edge-arrow");
Object.defineProperty(exports, "EdgeArrowProgram", { enumerable: true, get: function () { return __importDefault(edge_arrow_1).default; } });
var edge_arrow_head_1 = require("./programs/edge-arrow-head");
Object.defineProperty(exports, "EdgeArrowHeadProgram", { enumerable: true, get: function () { return __importDefault(edge_arrow_head_1).default; } });
var edge_clamped_1 = require("./programs/edge-clamped");
Object.defineProperty(exports, "EdgeClampedProgram", { enumerable: true, get: function () { return __importDefault(edge_clamped_1).default; } });
var edge_line_1 = require("./programs/edge-line");
Object.defineProperty(exports, "EdgeLineProgram", { enumerable: true, get: function () { return __importDefault(edge_line_1).default; } });
var edge_rectangle_1 = require("./programs/edge-rectangle");
Object.defineProperty(exports, "EdgeRectangleProgram", { enumerable: true, get: function () { return __importDefault(edge_rectangle_1).default; } });
var edge_triangle_1 = require("./programs/edge-triangle");
Object.defineProperty(exports, "EdgeTriangleProgram", { enumerable: true, get: function () { return __importDefault(edge_triangle_1).default; } });
